### PR TITLE
Sales Evidence Widget

### PR DESCRIPTION
**Pull request by johnEthicalTechnology**

### CODE REVIEW

File: console-export-2020-7-2_20-16-34.txt [hunk 1]
```diff
@@ -0,0 +1,149 @@
...Complex_Strata_Title_Not_Applicable":"Torrens","Retail_sqm":null,"Created_Time":"2020-06-02T10:21:18+10:00","Source_ID":null,"$editable":true,"Grade":null,"Longitude":"151.0922418","Postcode":"2138","Land_Tax":null,"Add_Note":null,"Lot":null,"State":"NSW","Property_Contacts":null,"User":null,"$review_process":{"approve":false,"reject":false,"resubmit":false},"Outgoings_Basis":null,"Council":"City of Canada Bay Council","Showroom_sqm":null,"$review":null,"Lead_Conversion_Time":null,"Overall_Sales_Duration":null,"Listing_No":"316002315","Property_Type_Portals":["Commercial","Industrial"],"Known_As":null,"Modified_Time":"2020-06-03T12:11:08+10:00","Prop_Zoning":"Zone IN1 \u2013 General Industrial","Sale_Type":[],"Land_Area_sqm":1769,"Car_Spaces":3,"$orchestration":false,"Sales_Cycle_Duration":null,"Property_Owners":null,"Sale_Date":"1984-06-07","Total_Outgoings":250000,"Reverse_Geocoded_Address":"2 Leeds St, Rhodes NSW 2138, Australia","$in_merge":false,"Insurance":null,"Warehouse_sqm":1000,"Tag":[],"Level_Suite":null,"Property_Category_Mailing":["Gym/fitness","Medical/dental","Automotive Service","Development","Office","Warehouse","Bulky Goods/Showroom"],"Managed":null,"Other":null})
```
Comment (jeznag) [inline console-export-2020-7-2_20-16-34.txt]: can we delete this?

File: src/components/DownloadSalesEvidenceListButton.tsx [hunk 2]
```diff
@@ -17,10 +17,11 @@ export default function DownloadSalesEvidenceListButton (props: DownloadButtonPr
...
 
-        let csvRow = `"${propertyAddress}","${landArea}","${buildArea}","${dateSold}","${salePrice}"\r\n`
-        csvRow = csvRow.replace(/null/g, '-')
-
-        return csvRow
+        if (landArea || buildArea || dateSold || salePrice) {
+            let csvRow = `"${propertyAddress}","${landArea}","${buildArea}","${dateSold}","${salePrice}"\r\n`
+            csvRow = csvRow.replace(/null/g, '-')
+            return csvRow
+        }
```
Comment (jeznag) [inline src/components/DownloadSalesEvidenceListButton.tsx:23]: csvRow will be undefined in the else clause. I'd suggest moving 

```
csvRow = csvRow.replace(/null/g, '-')
        return csvRow	            return csvRow
``` 
outside the if

File: src/components/ResultsTable.tsx [hunk 3]
```diff
@@ -1,42 +1,88 @@
+/* eslint-disable */
```
Comment (jeznag) [inline src/components/ResultsTable.tsx]: I'd rather not disable all eslint rules for a file.

use ```npx eslint --fix .``` to auto fix files

File: src/components/ResultsTable.tsx [hunk 4]
```diff
@@ -1,42 +1,88 @@
...
+                            <th>Address</th>
+                            <th>Land Area</th>
+                            <th>Build Area</th>
+                            <th>Date Sold</th>
+                            <th>Sale Price</th>
+                        </tr>
+                    </thead>
+                    <tbody>
+                        {props.results.map((result, index) => {
+                            let propertyAddress = result.Deal_Name
```
Comment (jeznag) [inline src/components/ResultsTable.tsx]: probably don't need to assign variables as there is no processing performed. Can direct reference them.

e.g. <td>{result.Deal_Name}</td>

File: src/components/SearchWidgetsWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+            <form>
+                <div className="radio-box">
+                    <label>
+                        <div className="radio">
+                            <input name='radio' type="radio" checked={props.filterInUse === 'BaseFilter'} onClick={() => props.setFilterInUse('BaseFilter')}/>
+                            <span className='radioName'>Map Widget</span>
+                            <input type="radio" checked={props.filterInUse === 'SalesEvidenceFilter'} onClick={() => props.setFilterInUse('SalesEvidenceFilter')}/>
+                            <span className='radioName'>Sales Evidence Widget</span>
+                            <input type="radio" />
```
Comment (jeznag) [inline src/components/SearchWidgetsWrapper.tsx]: perhaps disable this one for now

File: src/services/crmDataService.ts [hunk 6]
```diff
@@ -15,29 +15,32 @@ async function getPageOfProperties (pageNumber: number) {
     return response.data
 }
 
-const retrieveAllProperties = async function (pageNumber: number, retrievedProperties: UnprocessedResultsFromCRM[]): Promise<UnprocessedResultsFromCRM[]> {
-    const thisPageResults = await getPageOfProperties(pageNumber)
+const retrieveAllRecords = async function (pageNumber: number, retrievedProperties: UnprocessedResultsFromCRM[], zohoModuleToUse: string): Promise<UnprocessedResultsFromCRM[]> {
```
Comment (jeznag) [inline src/services/crmDataService.ts]: could change `retrievedProperties` to `retrievedRecords`

File: src/services/emptyObject [hunk 7]
```diff
@@ -0,0 +1,308 @@
+{
```
Comment (jeznag) [inline src/services/emptyObject]: are these files needed?

File: src/utils/filterResults.ts [hunk 8]
```diff
@@ -58,6 +61,10 @@ export default function filterResults (unsortedPropertyResults: UnprocessedResul
         const isUnderPropertyTypeLimit = matchTallies.propertyType < maxResultsForPropertyTypes
         const isUnderPropertyGroupLimit = matchTallies.propertyGroup < maxResultsForPropertyGroups
         const canAddAnotherProperty = isUnderNeighbourLimit || isUnderPropertyTypeLimit || isUnderPropertyGroupLimit
+        
+        if (filterInUse === 'SalesEvidenceFilter') {
+            canAddAnotherProperty = canAddAnotherProperty && salesEvidenceFilter(property, searchParameters)
```
Comment (jeznag) [inline src/utils/filterResults.ts]: seems like this is invalid JS (can't override a const)

File: src/utils/filterResults.ts [hunk 9]
```diff
@@ -86,9 +93,19 @@ export default function filterResults (unsortedPropertyResults: UnprocessedResul
                 if (!isDupeId) {
                     uniqueSearchRecords.push(property.id)
                 }
+                if (propertyTypeMatch) {
```
Comment (jeznag) [inline src/utils/filterResults.ts]: this increment logic seems to have been duplicated again (check line 81)

File: src/utils/salesEvidenceFilter.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
+import { SalesEvidenceFilterParams, SaleTypeEnum, UnprocessedResultsFromCRM, MinMaxNumberType, MinMaxDateType } from '../types'
+
+function numberFilter (property: UnprocessedResultsFromCRM, filterType: string, filterValues: MinMaxNumberType) {
+    return typeof property[filterType] === 'number' && property[filterType] >= filterValues.min && property[filterType] <= filterValues.max
+}
+
+function formatDateToString (date: Date): string {
```
Comment (jeznag) [inline src/utils/salesEvidenceFilter.ts]: would suggest using date.toISOString()

```
return date.toISOString().split('T')[0]
```

File: src/utils/salesEvidenceFilter.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    let dateDate: string | number = date.getDate()
+    const numberToLeftPadZero = 9
+    if (dateMonth <= numberToLeftPadZero) dateMonth = `0${dateMonth}`
+    if (dateDate <= numberToLeftPadZero) dateDate = `0${dateDate}`
+    return `${dateYear}-${dateMonth}-${dateDate}`
+}
+
+function dateFilter (property: UnprocessedResultsFromCRM, dateSold: MinMaxDateType): boolean {
+    if (typeof dateSold.min !== 'undefined' && typeof dateSold.max !== 'undefined') {
+        const minDate = formatDateToString(dateSold.min)
```
Comment (jeznag) [inline src/utils/salesEvidenceFilter.ts]: conversion to string is not necessary. Date objects respect > and <
https://www.w3resource.com/javascript-exercises/javascript-date-exercise-5.php

File: src/utils/salesEvidenceFilter.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+export default function salesEvidenceFilter (property: UnprocessedResultsFromCRM, filterParameters: SalesEvidenceFilterParams[]): boolean {
+    const {
+        landArea,
+        buildArea,
+        salePrice,
+        saleType,
+        dateSold
+    } = filterParameters[0]
```
Comment (jeznag) [inline src/utils/salesEvidenceFilter.ts]: why item 0? Shouldn't this support adding multiple filters?

File: src/utils/salesEvidenceFilter.ts [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    const FILTER_NOT_USED_NUM_TYPE = -1
```
Comment (jeznag) [inline src/utils/salesEvidenceFilter.ts]: Suggest BLANK_FILTER_VALUE = -1 as it's not a type

File: src/utils/salesEvidenceFilter.ts [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    const isLandAreaFilterInUse = landArea.min === FILTER_NOT_USED_NUM_TYPE && landArea.max === FILTER_NOT_USED_NUM_TYPE
+    const isInLandAreaRange = !isLandAreaFilterInUse && numberFilter(property, 'Land_Area_sqm', landArea)
+
+    const isBuildAreaFilterInUse = buildArea.min === FILTER_NOT_USED_NUM_TYPE && buildArea.max === FILTER_NOT_USED_NUM_TYPE
+    const isInBuildAreaRange = !isBuildAreaFilterInUse && numberFilter(property, 'Build_Area_sqm', buildArea)
+
+    const isSalePriceFilterInUse = salePrice.min === FILTER_NOT_USED_NUM_TYPE && salePrice.max === FILTER_NOT_USED_NUM_TYPE
+    const isInSalePriceRange = !isSalePriceFilterInUse && numberFilter(property, 'Sale_Price', salePrice)
+
+    const FILTER_NOT_USED_ARR_TYPE = 0
```
Comment (jeznag) [inline src/utils/salesEvidenceFilter.ts]: I don't think this is useful. ```saleType.length === 0``` is idiomatic javascript. No constant necessary

**Review by jeznag — CHANGES_REQUESTED** [review]

File: src/components/DownloadSalesEvidenceListButton.tsx (hunk 2 above)
Comment (johnEthicalTechnology) [reply src/components/DownloadSalesEvidenceListButton.tsx:23]: What do you mean? There is no else clause.

File: src/utils/salesEvidenceFilter.ts (hunk 12 above)
Comment (johnEthicalTechnology) [reply src/utils/salesEvidenceFilter.ts]: Nope, as per Chris's specs. Check the attached video in the July 1st email.

File: src/components/DownloadSalesEvidenceListButton.tsx (hunk 2 above)
Comment (jeznag) [reply src/components/DownloadSalesEvidenceListButton.tsx:23]: the implicit else clause.

```
if (landArea || buildArea || dateSold || salePrice) {
            csvRow = `"${propertyAddress}","${landArea}","${buildArea}","${dateSold}","${salePrice}"\r\n`
            csvRow = csvRow.replace(/null/g, '-')
} else { 
// csvRow is undefined
}
```

File: src/components/ResultsTable.tsx [hunk 15]
```diff
@@ -4,39 +4,147 @@ import { UnprocessedResultsFromCRM } from '../types'
...
-                                <td>{contactData?.Name || 'Contact Is Not Found'}</td>
-                            </tr>
-                        )
-                    })}
-                </tbody>
-            </table>
+        <div>
+            {props.filterInUse === 'BaseFilter' &&
+             (
+<<<<<<< Updated upstream
```
Comment (jeznag) [inline src/components/ResultsTable.tsx]: merge conflict

File: src/utils/salesEvidenceFilter.ts (hunk 12 above)
Comment (jeznag) [reply src/utils/salesEvidenceFilter.ts]: In that case I'd recommend not passing an array in. Code is confusing otherwise. If you don't want to do that, please add a comment.

File: src/services/informedObject [hunk 16]
```diff
@@ -0,0 +1,309 @@
+Pure object
```
Comment (jeznag) [inline src/services/informedObject]: can this be deleted?

File: src/services/emptyObject (hunk 7 above)
Comment (jeznag) [reply src/services/emptyObject]: File not removed

File: src/components/DownloadContactListButton.tsx [hunk 17]
```diff
@@ -8,22 +8,29 @@ type DownloadButtonProps = {
...
             const mobileNumbers = result.owner_details.map((owner: OwnerType) => owner.Mobile).filter((Mobile: string) => Mobile)
             const mobile = mobileNumbers.length > 0 ? mobileNumbers[0] : null
             const workPhones = result.owner_details.map((owner: OwnerType) => owner.Work_Phone).filter((Work_Phone: string) => Work_Phone)
             const workPhone = workPhones.length > 0 ? workPhones[0] : null
             const propertyAddress = result.Deal_Name
-            const propertyTypeMarketing = result.Property_Category_Mailing
-            const ownerData = result.owner_details.find((owner: OwnerType) => owner.Contact_Type === 'Owner')
-            const contactData = result.owner_details.find((owner: OwnerType) => owner.Contact_Type === 'Director')
+            const ownerDetails = result.owner_details
+            const contact: OwnerType = ownerDetails[0]
```
Comment (jeznag) [inline src/components/DownloadContactListButton.tsx]: I don't think this logic is correct. Why was it changed from ```const ownerData = result.owner_details.find((owner: OwnerType) => owner.Contact_Type === 'Owner')```

File: src/components/DownloadMailingListButton.tsx [hunk 18]
```diff
@@ -17,31 +16,40 @@ export function DownloadMailingListButton (props: DownloadButtonProps) {
         let email
         const propertyAddress = propertyObject.Deal_Name
         const ownerDetails = propertyObject.owner_details
-        const contact: OwnerType = ownerDetails[0]
-        let owner: OwnerType = ownerDetails[1]
+        const relatedContact: OwnerType = ownerDetails[0]
```
Comment (jeznag) [inline src/components/DownloadMailingListButton.tsx]: how can you know that it is always index 0? There may be more than 2 owners

File: src/components/DownloadMailingListButton.tsx [hunk 19]
```diff
@@ -17,31 +16,40 @@ export function DownloadMailingListButton (props: DownloadButtonProps) {
...
+            ownerOrRelatedContact = relatedContact
         }
 
+        const isDupe = ownerContactDupeRemoval.includes(`${ownerOrRelatedContact.Postal_Address}-${ownerOrRelatedContact.Name}`)
         if (!doNotMail || !returnToSender || !email) {
-            owner = owner ? (contact.Postal_Address ? contact : owner) : contact
-            const lastMailed = owner?.Last_Mailed || 'Last mailed has not been found'
-            csvRow = `"${owner?.Name}","${owner?.Contact_Type}","${postalAddress}","${owner?.Postal_Suburb}","${owner?.Postal_State}","${owner?.Postal_Postcode}","${propertyAddress}, ${lastMailed}\r\n`
-            csvRow = csvRow.replace(/null/g, '-')
+            const checker = null || undefined
```
Comment (jeznag) [inline src/components/DownloadMailingListButton.tsx]: I'm not sure this does what you want it to do.

null || undefined === undefined

What is your desired logic?

File: src/components/DownloadSalesEvidenceListButton.tsx (hunk 2 above)
Comment (jeznag) [reply src/components/DownloadSalesEvidenceListButton.tsx:23]: You didn't make any changes. I suggest you move
csvRow = csvRow.replace(/null/g, '-')

outside the if

File: src/components/ResultsTable.tsx [hunk 20]
```diff
@@ -4,39 +4,83 @@ import { UnprocessedResultsFromCRM } from '../types'
...
+                                 <th>Address</th>
+                                 <th>Land Area</th>
+                                 <th>Build Area</th>
+                                 <th>Date Sold</th>
+                                 <th>Sale Price</th>
+                             </tr>
+                         </thead>
+                         <tbody>
+                             {props.results.map((result, index) => {
+                                 const propertyAddress = result.Deal_Name
```
Comment (jeznag) [inline src/components/ResultsTable.tsx]: no reason to create variables here

File: src/components/salesEvidenceSearch.tsx [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+    searchParameters: IntersectedSearchAndFilterParams
+    changeSearchParameters: (newParameters: IntersectedSearchAndFilterParams) => void
+}
+
+export function SalesEvidenceSearchWidget (props: SearchWidgetProps) {
+    return (
+        <>
+            <SearchWidget changeSearchParameters={props.changeSearchParameters} searchParameters={props.searchParameters} />
+            <form className="wrapper">
+                <label className="eight">Land Area m2<br />
```
Comment (jeznag) [inline src/components/salesEvidenceSearch.tsx]: 2 should be superscript. You can use the <sup> tag

File: src/utils/filterResults.ts [hunk 22]
```diff
@@ -37,13 +38,37 @@ function getOwnerData (property: UnprocessedResultsFromCRM) {
     return ownerData
 }
 
-export default function filterResults (unsortedPropertyResults: UnprocessedResultsFromCRM[], searchParameters: SearchParametersType[]): { matchedProperties: UnprocessedResultsFromCRM[], uniqueSearchRecords: string[] } {
+export default function filterResults (unsortedPropertyResults: UnprocessedResultsFromCRM[], searchParameters: IntersectedSearchAndFilterParams[], filterInUse: string): { matchedProperties: UnprocessedResultsFromCRM[], uniqueSearchRecords: string[] } {
+    let desiredPropertyTypes = searchParameters[0].propertyTypes
```
Comment (jeznag) [inline src/utils/filterResults.ts]: seems like this would break the default filter. It should not be referencing index 0

File: src/utils/filterResults.ts [hunk 23]
```diff
@@ -37,13 +38,37 @@ function getOwnerData (property: UnprocessedResultsFromCRM) {
...
+    let desiredPropertyGroups = searchParameters[0].propertyGroups
+    let isPropertyTypeFilterInUse = desiredPropertyTypes.length !== 0
```
Comment (jeznag) [inline src/utils/filterResults.ts]: why is this needed? 

`isUnderPropertyTypeLimit && matchForPropertyTypes(property, desiredPropertyTypes)`

would always be false if desiredPropertyTypes is []

File: src/components/DownloadContactListButton.tsx (hunk 17 above)
Comment (johnEthicalTechnology) [reply src/components/DownloadContactListButton.tsx]: I'm not sure. @haydenf why the change?

File: src/components/DownloadMailingListButton.tsx (hunk 18 above)
Comment (johnEthicalTechnology) [reply src/components/DownloadMailingListButton.tsx]: @haydenf?

File: src/components/DownloadContactListButton.tsx (hunk 17 above)
Comment (haydenf) [reply src/components/DownloadContactListButton.tsx]: Because not every related contact is of contact type director, where related contact seems to always be ownerdetails[0] and owner is ownerdetails[1]

File: src/components/DownloadMailingListButton.tsx (hunk 18 above)
Comment (haydenf) [reply src/components/DownloadMailingListButton.tsx]: I didn't think of this, do properties have multiple owners? If so I'm unsure of how to go about a fix for it

File: src/components/DownloadMailingListButton.tsx (hunk 19 above)
Comment (haydenf) [reply src/components/DownloadMailingListButton.tsx]: I'm also confused about this, I just want to make sure field wasn't empty but I did it in a weird way, fixed in next push

File: src/components/ResultsTable.tsx (hunk 20 above)
Comment (haydenf) [reply src/components/ResultsTable.tsx]: Fixed

File: src/components/salesEvidenceSearch.tsx (hunk 21 above)
Comment (haydenf) [reply src/components/salesEvidenceSearch.tsx]: Fixed

File: src/components/DownloadMailingListButton.tsx (hunk 18 above)
Comment (jeznag) [reply src/components/DownloadMailingListButton.tsx]: I'd suggest using the code I had in other places where it uses ownerDetails.find to locate

File: src/components/DownloadContactListButton.tsx (hunk 17 above)
Comment (jeznag) [reply src/components/DownloadContactListButton.tsx]: yeah that's why we should use const ownerData = result.owner_details.find((owner: OwnerType) => owner.Contact_Type === 'Owner')

We cannot assume that index 0 is always the related contact

File: src/utils/filterResults.ts (hunk 23 above)
Comment (johnEthicalTechnology) [reply src/utils/filterResults.ts]: According to Chris's spec he wants:
1) Default: nothing entered into anywhere >> return all records
2) If either property type or property group is used >> disable the one not used and return properties that have the value of the type or group entered. They always use OR logic even if used together.
3) If property managed is used with any combination of type or group >> return the combination and if the property is managed. The managed is uses AND logic with group or type.

I've made the default state for group and type to an empty array so I know they don't want to use that filter and it disables it by making max results zero. But if none of them are used them then I'm assuming they want to return all properties that have any group or type value.

Comment (jeznag) [reply src/utils/filterResults.ts]: I don't think you need a variable for isPropertyTypeFilterInUse or any of the extra if statements that were added.

I think they already worked independently the way we had it before. 

For example, if the property type filter is disabled, 

```isUnderPropertyTypeLimit && matchForPropertyTypes(property, desiredPropertyTypes)``` will always return false because `desiredPropertyTypes` would be `[]` and `matchForPropertyTypes` would also return false because 

image


RE default state: shouldn't the default state be All for all filters?

File: src/components/DownloadContactListButton.tsx (hunk 17 above)
Comment (jeznag) [reply src/components/DownloadContactListButton.tsx]: This is still an issue

File: src/components/DownloadMailingListButton.tsx (hunk 18 above)
Comment (jeznag) [reply src/components/DownloadMailingListButton.tsx]: Still an issue

File: src/components/DownloadMailingListButton.tsx (hunk 19 above)
Comment (jeznag) [reply src/components/DownloadMailingListButton.tsx]: Still an issue

File: src/components/DownloadMailingListButton.tsx [hunk 24]
```diff
@@ -1,47 +1,53 @@
...
         let email
         const propertyAddress = propertyObject.Deal_Name
-        const ownerDetails = propertyObject.owner_details
-        const contact: OwnerType = ownerDetails[0]
-        let owner: OwnerType = ownerDetails[1]
+        const relatedContact = propertyObject.owner_details?.find((owner: OwnerType) => owner.Contact_Type === 'Director')
+        const owner = propertyObject.owner_details?.find((owner: OwnerType) => owner.Contact_Type === 'Owner')
 
-        if (!contact && owner) {
+        let ownerOrRelatedContact = owner || relatedContact || null
```
Comment (jeznag) [inline src/components/DownloadMailingListButton.tsx:21]: For future reference || null may not be need as it would default to undefined

File: src/components/ResultsTable.tsx [hunk 25]
```diff
@@ -4,39 +4,78 @@ import { UnprocessedResultsFromCRM } from '../types'
...
+                                 <th>Property Address</th>
+                                 <th>Owner</th>
+                                 <th>Contact</th>
+                             </tr>
+                         </thead>
+                         <tbody>
+                             {props.results.map((result, index) => {
+                                 let propertyAddress = result.Deal_Name
+                                 if (!result.Latitude || !result.Longitude) {
+                                     propertyAddress = `${result.Deal_Name} - Geocordinates N/A, cannot display on map.`
```
Comment (jeznag) [inline src/components/ResultsTable.tsx]: nitpick: coordinates spelt incorrectly - can fix it up later

File: src/components/salesEvidenceSearch.tsx [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+                </label>
+                <SaleTypeDropdown chosenSaleType={props.searchParameters.saleType} changeSaleTypes={(newSaleTypes) => {
+                    props.changeSearchParameters({
+                        ...props.searchParameters,
+                        saleType: newSaleTypes
+                    })
+                }} />
+                <label className="thirteen">Select All Records for Sales Evidence Sub Filter
+                    <input type="checkbox" onChange={(e: ChangeEvent<HTMLInputElement>) => {
+                        const allRecords = e.target.value === 'on'
```
Comment (jeznag) [inline src/components/salesEvidenceSearch.tsx]: where does value on come from?

File: src/utils/emailAndIdExtract.ts [hunk 27]
```diff
@@ -9,7 +9,7 @@ export default function emailAndIdExtract (results: UnprocessedResultsFromCRM[])
     const propertyObjects = results.slice(1)
 
     const emailsAndIds = propertyObjects.reduce((resultsArray: MassMailObject[], property: UnprocessedResultsFromCRM) => {
-        if (typeof property.owner_details[0].Email === 'string') {
+        if (typeof property.owner_details[0]?.Email === 'string') {
```
Comment (jeznag) [inline src/utils/emailAndIdExtract.ts]: probably want to change this to not be index based

File: src/utils/filterResults.ts [hunk 28]
```diff
@@ -37,58 +38,74 @@ function getOwnerData (property: UnprocessedResultsFromCRM) {
...
+                const shouldAddProperty = isManaged && canAddBasedOnFilters
+
+                if (shouldAddProperty) {
+                    if (ownerData.length > 0) {
+                        property.owner_details = ownerData
+                    }
+                    if (propertyTypeMatch) {
+                        matchTallies.propertyType += 1
+                    }
+                    if (propertyGroupMatch) {
```
Comment (jeznag) [inline src/utils/filterResults.ts:93]: Can you double check this logic? I thought it was only meant to count a property group match if it was not a property type match? Maybe things changed though.

File: src/utils/salesEvidenceFilter.ts [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+
+function numberFilter (filterType: string, filterValues: MinMaxNumberType, property: UnprocessedResultsFromCRM) {
+    return typeof property[filterType] === 'number' && property[filterType] >= filterValues.min && property[filterType] <= filterValues.max
+}
+
+function dateFilter (dateSold: MinMaxDateType, property: UnprocessedResultsFromCRM): boolean {
+    if (typeof dateSold.min !== 'undefined' && typeof dateSold.max !== 'undefined') {
+        const minDate = dateSold.min.toISOString().split('T')[0]
+        const maxDate = dateSold.max.toISOString().split('T')[0]
+        return !property.Sale_Date && property.Sale_Date >= minDate && property.Sale_Date <= maxDate
```
Comment (jeznag) [inline src/utils/salesEvidenceFilter.ts]: I don't understand this. Seems like this would never evaluate to true. ```!property.Sale_Date && property.Sale_Date >= minDate```

Should it be
```
property.Sale_Date && property.Sale_Date >= minDate
```

File: src/utils/salesEvidenceFilter.ts [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+export default function salesEvidenceFilter (filterParameters: SalesEvidenceFilterParams, property: UnprocessedResultsFromCRM): boolean {
+    const {
+        landArea,
+        buildArea,
+        salePrice,
+        saleType,
+        dateSold
+    } = filterParameters
+    const BLANK_FILTER_VALUE = -1
+    const isLandAreaFilterInUse = landArea.min === BLANK_FILTER_VALUE && landArea.max === BLANK_FILTER_VALUE
```
Comment (jeznag) [inline src/utils/salesEvidenceFilter.ts]: Seems wrong logic. This means the filter is not in use

File: src/components/salesEvidenceSearch.tsx (hunk 26 above)
Comment (johnEthicalTechnology) [reply src/components/salesEvidenceSearch.tsx]: `e.target.value` is equal to 'on'.

File: src/utils/salesEvidenceFilter.ts [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+    const isInLandAreaRange = !isLandAreaFilterInUse && numberFilter('Land_Area_sqm', landArea, property)
```
Comment (jeznag) [inline src/utils/salesEvidenceFilter.ts]: Inverse logic. Should be isLandAreaFilterInUse not ! isLandAreaFilterInUse

**Review by jeznag — CHANGES_REQUESTED** [review]
Few little things

File: src/utils/filterResults.ts (hunk 28 above)
Comment (johnEthicalTechnology) [reply src/utils/filterResults.ts:93]: I put an image for the requirements of the base filter in the Zoho chat.

File: src/utils/salesEvidenceFilter.ts (hunk 29 above)
Comment (johnEthicalTechnology) [reply src/utils/salesEvidenceFilter.ts]: I'm going to change it to `!!property.Sale_Date` b/c i'll get this type error: `Type 'boolean | ""' is not assignable to type 'boolean'. Type '""' is not assignable to type 'boolean'.ts(2322)`.

File: src/utils/salesEvidenceFilter.ts (hunk 31 above)
Comment (johnEthicalTechnology) [reply src/utils/salesEvidenceFilter.ts]: I think the variable is not correctly named. When it's `true` it means the filter is not in use b/c it is equal to `-1` which is the default value of the field. It is only when this is equal to `false` when the filter is in use which is why I have bang operator being used here.

**Review by jeznag — APPROVED** [review]
